diff --git a/CONFIG b/CONFIG
index ddf8341a5..290db8d46 100644
--- a/CONFIG
+++ b/CONFIG
@@ -10,9 +10,6 @@
 # Installation prefix
 CONFIG_PREFIX="/usr/local"
 
-# Destination directory for the Python libraries
-CONFIG_PYDIR=
-
 # Target architecture
 CONFIG_ARCH=native
 
diff --git a/Makefile b/Makefile
index 3aeae41ad..e91d44caa 100644
--- a/Makefile
+++ b/Makefile
@@ -24,7 +24,6 @@ DIRS-$(CONFIG_VFIO_USER) += vfiouserbuild
 DIRS-$(CONFIG_SMA) += proto
 DIRS-$(CONFIG_XNVME) += xnvmebuild
 DIRS-$(CONFIG_GOLANG) += go/rpc
-DIRS-y += python
 
 .PHONY: all clean $(DIRS-y) include/spdk/config.h mk/config.mk \
 	cc_version cxx_version .libs_only_other .ldflags ldflags install \
diff --git a/configure b/configure
index 0428e5cea..b66d3be71 100755
--- a/configure
+++ b/configure
@@ -28,7 +28,6 @@ function usage() {
 	echo " --cross-prefix=prefix     Prefix for cross compilation (default: none)"
 	echo "                           example: aarch64-linux-gnu"
 	echo " --libdir=path             Configure installation path for the libraries (default: \$prefix/lib)"
-	echo " --pydir=path              Configure installation path for the python libraries"
 	echo " --max-lcores=VAL          DPDK configuration. VAL defines maximum number of lcores supported"
 	echo "                           by EAL, or enables autodetection if set to 'detect'. When 'detect'"
 	echo "                           is specified, DPDK will detect number of cores in the system during"
@@ -301,9 +300,6 @@ for i in "$@"; do
 		--libdir=*)
 			CONFIG[LIBDIR]="${i#*=}"
 			;;
-		--pydir=*)
-			CONFIG[PYDIR]="${i#*=}"
-			;;
 		--enable-debug)
 			CONFIG[DEBUG]=y
 			;;
diff --git a/proto/Makefile b/proto/Makefile
index 310c4fe61..767b93bd3 100644
--- a/proto/Makefile
+++ b/proto/Makefile
@@ -16,6 +16,8 @@ all: $(protopy)
 clean:
 	$(Q)$(RM) $(protopy)
 
+# TODO: we should probably write a proper install rule here instead of just blindly copying all
+# python scripts when building the RPMs
 install:
 uninstall:
 
diff --git a/rpmbuild/spdk.spec b/rpmbuild/spdk.spec
index bf9a1fb5d..4ac8bc236 100644
--- a/rpmbuild/spdk.spec
+++ b/rpmbuild/spdk.spec
@@ -110,6 +110,7 @@ mkdir -p %{buildroot}/usr/libexec/spdk
 mkdir -p %{buildroot}/etc/bash_completion.d
 mkdir -p %{buildroot}/etc/profile.d
 mkdir -p %{buildroot}/etc/ld.so.conf.d
+mkdir -p %{buildroot}%{python3_sitelib}
 
 cat <<-EOF > %{buildroot}/etc/ld.so.conf.d/spdk.conf
 %{libdir}
@@ -125,6 +126,7 @@ export PATH
 EOF
 
 cfs %{buildroot}/usr/libexec/spdk scripts
+cfs  %{buildroot}%{python3_sitelib} python/spdk
 ln -s /usr/libexec/spdk/scripts/bash-completion/spdk %{buildroot}/etc/bash_completion.d/
 
 # We need to take into the account the fact that most of the scripts depend on being
@@ -140,7 +142,7 @@ ln -s /usr/local/include %{buildroot}/usr/libexec/spdk
 /etc/bash_completion.d/*
 /usr/libexec/spdk/*
 /usr/local/bin/*
-%{python3_sitelib}/spdk*/*
+%{python3_sitelib}/spdk/*
 
 
 %package devel
diff --git a/test/common/autobuild_common.sh b/test/common/autobuild_common.sh
index 10d9ded14..5e0f724c9 100755
--- a/test/common/autobuild_common.sh
+++ b/test/common/autobuild_common.sh
@@ -308,7 +308,7 @@ test_make_uninstall() {
 	# Create empty file to check if it is not deleted by target uninstall
 	touch "$SPDK_WORKSPACE/usr/lib/sample_xyz.a"
 	$MAKE $MAKEFLAGS uninstall DESTDIR="$SPDK_WORKSPACE" prefix=/usr
-	if [[ $(find "$SPDK_WORKSPACE/usr" -maxdepth 1 -mindepth 1 | wc -l) -ne 3 ]] || [[ $(find "$SPDK_WORKSPACE/usr/lib/" -maxdepth 1 -mindepth 1 | wc -l) -ne 1 ]]; then
+	if [[ $(find "$SPDK_WORKSPACE/usr" -maxdepth 1 -mindepth 1 | wc -l) -ne 2 ]] || [[ $(find "$SPDK_WORKSPACE/usr/lib/" -maxdepth 1 -mindepth 1 | wc -l) -ne 1 ]]; then
 		ls -lR "$SPDK_WORKSPACE"
 		echo "Make uninstall failed"
 		exit 1
